api/src/middleware/rate-limiter.js

```diff
@@ -0,0 +1,42 @@
+const auth = require('../auth');
+const rateLimitService = require('../services/rate-limit');
+const serverUtils = require('../server-utils');
+
+const getBasicAuthUsername = (req) => auth.basicAuthCredentials(req);
+
+const consume = (req, res) => {
+  res.on('finish', () => {
+    if (res.statusCode === 401 || res.statusCode === 429) {
+      rateLimitService.consume(getKeys(req));
+    }
+  });
+};
+
+const getKeys = (req) => {
+  const keys = [
+    req.ip,
+    req.body?.user,
+    req.body?.password
+  ];
+
+  const basicAuth = getBasicAuthUsername(req);
+  if (basicAuth) {
+    keys.push(basicAuth.username);
+    keys.push(basicAuth.password);
+  }
+
+  return keys;
+};
```

Comment:
The keys I've gone with are
- requesting IP address so a single computer can't spam login attempts,
- the username so if there is a target user or known user the attacker can't spam all combinations of possible passwords, and
- the password so the attacker can't spam common passwords against all existing users.

This is still vulnerable to distributed attacks, for a range of user/pass combinations.